tmt/steps/prepare/artifact/providers/repository.py

```diff
@@ -1,126 +1,188 @@
-
-        sudo = "sudo " if not guest.facts.is_superuser else ""
-        self.logger.info(f"Installing repository '{url}' to '{repo_dest}'.")
-
-        try:
-            # TODO: Add retry Mechanism
-            guest.execute(
-                ShellScript(f"{sudo}curl -L --fail -o {quote(str(repo_dest))} {quote(url)}"),
-                silent=True,
-            )
-        except GeneralError as error:
-            raise DownloadError(f"Failed to download repository file to '{repo_dest}'.") from error
-
-        # 2. Populate the RPM list for discovery purposes
-        self._fetch_rpms(guest, repo_dest)
-
-        self.logger.info("Repository setup is complete.")
-        # 3. Return list of available Artifacts
-        # TODO: Finalize contract of what needs to be returned from Artifact Repository
+        self.repository = Repository.from_url(url=self.id, logger=self.logger)
+
+        # Install the repository using the guest's package manager
+        guest.package_manager.install_repository(self.repository)
+
+        # Load the artifacts using list_packages
+        package_list = guest.package_manager.list_packages(self.repository)
+
+        # Initialize the list before populating
+        self._artifact_list = []
+
+        for pkg in package_list:
+            try:
+                # Use the new utility function to parse the package string
+                raw_artifact = {**parse_rpm_string(pkg_string=pkg), "url": self.id}
+                self._artifact_list.append(RpmArtifactInfo(_raw_artifact=raw_artifact))
+
+            except ValueError:
+                # Catches both regex failing to match (ValueError)
+                # or an explicit ValueError raised by the utility function.
+                self.logger.warning(f"Failed to parse malformed package string '{pkg}'. Skipping.")
+                continue
+
+            except Exception as error:
+                # Catch any other unexpected errors
+                self.logger.debug(f"Unexpected error while parsing package '{pkg}': {error}.")
+                continue
+
+        self.logger.debug(f"Successfully discovered '{len(self._artifact_list)}' artifacts.")
+
         return []
+
+
+# FIXME: Make this function more robust. The current regex-based parsing
+# is a "happy path" implementation and will fail on complex or
+# unusually-named packages. This is acceptable for now but should
+# be hardened later
+# Regex to parse N-E:V-R.A format.
+# Groups: 1:Name, 3:Epoch (optional), 4:Version, 5:Release, 6:Arch
+_PKG_REGEX = re.compile(
+    r"^(?P<name>.+?)"  # 1: Name (non-greedy)
+    r"-"  # (literal hyphen)
+    r"((?P<epoch>\d+):)?"  # 2, 3: Epoch (optional group)
+    # 4: Version (must contain at least one digit)
+    r"(?P<version>[^-:]*\d[^-:]*)"
+    r"-"  # (literal hyphen)
+    r"(?P<release>[^-]+)"  # 5: Release (no hyphens)
+    r"\."  # (literal dot)
+    r"(?P<arch>[^.]+)$"  # 6: Arch
+)
+
+
+def parse_rpm_string(pkg_string: str) -> dict[str, str]:
```

Comment:
Another point for https://github.com/teemtee/tmt/issues/4221 - investigate how to make the "rpm" more hidden, as it's just one kind of repositories out there. We may end up with some `RpmRepository`, or `YumRepository` implementations of a general `Repository` abstract class later. Let's just track it for now.